scripts/generate_can.py

```diff
@@ -0,0 +1,258 @@
+    value = re.sub(r'[^a-zA-Z0-9]', '_', value)
+    return value
+
+
+# load yaml file
+with open(YAML_FILE_PATH) as file:
+    can_yaml = yaml.load(file, Loader=yaml.FullLoader)
+
+bus_name = list(can_yaml.keys())[0]
+bus_frequency = can_yaml[bus_name]['bus_frequency']
+canid_filter_mask = can_yaml[bus_name]['canid_filter_mask']
+canid_filters = can_yaml[bus_name]['canid_filters']
+
+# sort nodes in alphabetical order
+can_yaml[bus_name]['nodes'].sort()
+
+roboteq_enums = can_yaml[bus_name]['roboteq_canids']
+
+# extract CAN nodes
+nodes = []
+for node in can_yaml[bus_name]['nodes']:
+    nodes.append(
+        cantools.database.can.Node(name=node, comment=None)
+    )
+
+# sort messages by id
+can_yaml[bus_name]['messages'].sort(key=lambda x: x[list(x.keys())[0]]['id'])
+
+# extract CAN messages
+messages = []
+for message in can_yaml[bus_name]['messages']:
+    message_name = list(message.keys())[0]
+    message = message[message_name]
+
+    # add to autogen dicts
+    msg_snake_upper = camel_to_snake_case(message_name).upper()
+    AUTOGEN_message_enums[msg_snake_upper] = message['id']
+    AUTOGEN_msg_map[msg_snake_upper] = []
+    AUTOGEN_original_msg_names[msg_snake_upper] = message_name
+
+    # sort senders and receivers
+    message['senders'].sort()
+    message['receivers'].sort()
+
+    # sort signals by startbit
+    message['signals'].sort(key=lambda x: x[list(x.keys())[0]]['startbit'])
+
+    # extract message signals
+    signals = []
+    for signal in message['signals']:
+        signal_name = list(signal.keys())[0]
+        signal = signal[signal_name]
+
+        # add to autogen signal enums list and msg map dict
+        signal_snake_upper = camel_to_snake_case(signal_name).upper()
+        if 'values' in signal and signal['values'] is not None:
+            AUTOGEN_signal_enum_choices[signal_snake_upper] = signal['values']
+        AUTOGEN_signal_enums.append(signal_snake_upper)
+        AUTOGEN_msg_map[msg_snake_upper].append(signal_snake_upper)
+
+        length = signal['length']
+        is_signed = signal['is_signed']
+
+        # NOTE: it is assumed that min/max bounds do not cover SNA values
+
+        # if given both scale/offset and min/max
+        if 'scale' in signal and 'offset' in signal and 'min' in signal and 'max' in signal:
+            scale = signal['scale']
+            offset = signal['offset']
+            sig_min = signal['min']
+            sig_max = signal['max']
+
+        # else if only given scale/offset (no min/max)
+        elif 'scale' in signal and 'offset' in signal:
+            scale = signal['scale']
+            offset = signal['offset']
+
+            # calculate min/max based on scale/offset and length
+            sig_min = - (2**(length-1)-1) * scale + \
+                offset if is_signed else offset
+            sig_max = (2**(length-1)-1) * scale + \
+                offset if is_signed else (2**length-2) * scale + offset
```

Comment:
This conversion calculation from min/max to scale/offset makes it so that the range does not include the SNA value (currently set to largest possible value representable by the bits). I realized that a caveat of this is that the `signal_in_range()` check will fail if the signal value is SNA. Does this approach make sense?